Support nested interrupts on AArch32 (non-Cortex-M)

stephanosio opened the pull request:
```
arch: arm: aarch32: Rework non-Cortex-M context preservation 

The current context preservation implementation saves the spsr and
lr_irq registers, which contain the cpsr and pc register values of the
interrupted context, in the thread callee-saved block and this prevents
nesting of interrupts because these values are required to be part of
the exception stack frame to preserve the nested interrupt context.

This commit reworks the AArch32 non-Cortex-M context preservation
implementation to save the spsr and lr_irq registers in the exception
stack frame to allow preservation of the nested interrupt context as
well as the interrupted thread context.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

Includes the commits from #23518 and #23614.

Closes #22670.

stephanosio committed:
tests: kernel: interrupt: Enable nested interrupt test for Cortex-R

This commit enables nested interrupt test for the Cortex-R platforms
that use the ARM Generic Interrupt Controller (GIC).

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
arch: arm: aarch32: Rework non-Cortex-M context preservation

The current context preservation implementation saves the spsr and
lr_irq registers, which contain the cpsr and pc register values of the
interrupted context, in the thread callee-saved block and this prevents
nesting of interrupts because these values are required to be part of
the exception stack frame to preserve the nested interrupt context.

This commit reworks the AArch32 non-Cortex-M context preservation
implementation to save the spsr and lr_irq registers in the exception
stack frame to allow preservation of the nested interrupt context as
well as the interrupted thread context.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: qemu_cortex_r5: Remove ignore tags for working tests

This commit removes the ignore tags for the tests that work after the
changes in the PR #23636.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
`tests/kernel/interrupt:test_nested_isr` Test Output
```
starting test - test_nested_isr
Triggering irq : 14
isr0: Enter
Triggering irq : 15
isr1: Enter
isr1: Leave
isr0: Leave
PASS - test_nested_isr
```

stephanosio commented:
I just realised I have a more recent (and advanced) version of this in one of my private branches ... (it's been a while since I worked on this and my memory is failing ...). I will push that version tomorrow.

stephanosio commented:
Shippable fails because `tests/arch/arm/arm_irq_vector_table/arch.interrupt.arm.irq_vector_table` is failing, but this test is not supposed to be run for `qemu_cortex_r5` because its `testcase.yaml` has `filter: CONFIG_ARMV6_M_ARMV8_M_BASELINE or CONFIG_ARMV7_M_ARMV8_M_MAINLINE`.

It looks like we have a bug in `sanitycheck`?

cc @nashif

SebastianBoe commented:
What do you mean by "The direct cause is $<TARGET_FILE:zephyr_final>". What is it evalating to, and what would you like it to evaluate to?

stephanosio commented:
> What do you mean by "The direct cause is $<TARGET_FILE:zephyr_final>". What is it evalating to, and what would you like it to evaluate to?

@SebastianBoe In summary, when `CONFIG_GEN_ISR_TABLES=n`, having `$<TARGET_FILE:zephyr_final>` in `board.cmake` results in the following error:
```
CMake Error at ../../../../cmake/emu/qemu.cmake:304 (add_custom_target):
  Error evaluating generator expression:

    $<TARGET_FILE:zephyr_final>

  No target "zephyr_final"
Call Stack (most recent call first):
  ../../../../CMakeLists.txt:1340 (include)
```

This is (or rather lack thereof) what is causing the error:
https://github.com/zephyrproject-rtos/zephyr/blob/b7f47f7805e5c6dbf92dc624a588f9a6b283258d/CMakeLists.txt#L729-L763

SebastianBoe commented:
Hi, okay, so the problem is that the target name for the final elf could be one of several.

This is usually not a problem, at line 10k of zephyr/CMakeLists.txt we set the variable logical_target_for_zephyr_elf to be whatever the correct target is for the current Kconfig.

And users usually de-reference logical_target_for_zephyr_elf to find the target name. Your problem is that you need to de-reference it in board.cmake, which is evaluated before line 10k of zephyr/CMakeLists.txt.

The solution is to defer evaluation using a generator expression. You need to define some target property that you can refer to from board.cmake, and then set later on, e.g. at line 10k of zephyr/CMakeListst.txt.

I'd do something like:

`$<TARGET_FILE:$<TARGET_PROPERTY:zephyr_property_target,logical_target_for_zephyr_elf>>
`
to refer to it from board.cmake

and

```
  set_property(TARGET zephyr_property_target
               PROPERTY logical_target_for_zephyr_elf
               ${logical_target_for_zephyr_elf})
```

to set it at line 10k in CMakeListst.txt.

The rules about what you can do from a generator expression are unpredictable, so this might not work, but it is the cleanest solution if it does.

stephanosio commented:
@SebastianBoe Thanks for looking into this. I tried your solution and that did work.

One question I have is: is there some kind of "delayed variable expansion" support in CMake? I think it would be the best if we can just do `set(QEMU_KERNEL_OPTION "...\$<TARGET_FILE:\${logical_target_for_zephyr_elf}>..."` in `board.cmake`, so that these get expanded/resolved in `cmake/emu/qemu.cmake`, where the vars are available.

There seem to be some "unofficial util functions" (or rather "hacks") that expand escaped vars, but I would prefer to use a CMake built-in function/macro if there is one? If not, should we add the util function to the Zephyr CMake codebase?

stephanosio commented:
> @SebastianBoe Thanks for looking into this. I tried your solution and that did work.
> 
> One question I have is: is there some kind of "delayed variable expansion" support in CMake? I think it would be the best if we can just do `set(QEMU_KERNEL_OPTION "...\$<TARGET_FILE:\${logical_target_for_zephyr_elf}>..."` in `board.cmake`, so that these get expanded/resolved in `cmake/emu/qemu.cmake`, where the vars are available.
> 
> There seem to be some "unofficial util functions" (or rather "hacks") that expand escaped vars, but I would prefer to use a CMake built-in function/macro if there is one? If not, should we add the util function to the Zephyr CMake codebase?

@SebastianBoe Actually, I think I found the solution:
```
...
elseif(DEFINED QEMU_KERNEL_OPTION)
  string(CONFIGURE "${QEMU_KERNEL_OPTION}" QEMU_KERNEL_OPTION)
```
This seems to do the trick.

stephanosio committed:
cmake: qemu: Support delayed expansion of QEMU_KERNEL_OPTION

This commit allows the `QEMU_KERNEL_OPTION` variable, which can be
overridden by the `board.cmake`, to contain references to the variables
that are not available at the time of `board.cmake` inclusion, by
expanding its escpaed variable references in `cmake/emu/qemu.cmake`
which is included nearby the end of the root `CMakeLists.txt`.

With this change, the `board.cmake` can escape variable references as
follows and allow them to be expanded later:

set(QEMU_KERNEL_OPTION "-device;loader,file=\$<TARGET_FILE:$\{...}>")

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

SebastianBoe commented on boards/arm/qemu_cortex_r5/board.cmake:
```diff
@@ -14,7 +14,7 @@ set(QEMU_FLAGS_${ARCH}
 
 set(QEMU_KERNEL_OPTION
-  "-device;loader,file=$<TARGET_FILE:zephyr_final>,cpu-num=4"
+  "-device;loader,file=\$<TARGET_FILE:\${logical_target_for_zephyr_elf}>,cpu-num=4"
```
Interesting, I can't recall seeing this before.

Another tool in the toolbox.

stephanosio commented:
Rebased

stephanosio commented:
Rebased

stephanosio commented:
Rebased

ioannisg commented on arch/arm/core/aarch32/exc_exit.S:
```diff
@@ -73,11 +74,17 @@ SECTION_SUBSEC_FUNC(TEXT, _HandlerModeExit, z_arm_int_exit)
 
 #ifdef CONFIG_PREEMPT_ENABLED
-    ldr r0, =_kernel
+    ldr r3, =_kernel
```
although this seems ok - why did you have to change the register here?

stephanosio replied:
mainly so that I can free up `r0` and `r1` for use later (see below this line) while keeping the `_kernel` pointer in `r3`.

ioannisg replied:
couldn't you keep kernel in r0 and do 

```
  ldr r3, [r0, #_kernel_offset_to_nested]
  cmp r3, #1
```

stephanosio replied:
I prefer to keep the values that are referenced multiple times in the upper registers (in terms of caller-saved, that is) and use the lower registers as scratch.

That seems to be the convention used throughout the Zephyr ARM arch code as well (we are loading `_kernel` into `r2` and using `r0` and `r1` as scratch registers most of the time).

ioannisg replied:
never thought there's actually a convention about it, and I am certainly did not following it myself :)

OK, I was mostly worried that I am missing something here, but if this is a matter of preference, then i won't object

ioannisg reviewed (commented):
Looks good - just a small questions around a scratch register

ioannisg commented:
cc @bbolen

ioannisg commented:
@carlocaione pls, review
